File: app/src/main/java/com/platdmit/simplecloudmanager/domain/converters/implement/ServerConvertImp.java
package com.platdmit.simplecloudmanager.domain.converters.implement;

import com.platdmit.simplecloudmanager.data.api.models.ApiServer;
import com.platdmit.simplecloudmanager.data.database.entity.DbServer;
import com.platdmit.simplecloudmanager.domain.converters.ServerConverter;
import com.platdmit.simplecloudmanager.domain.models.Server;

import org.joda.time.DateTime;
import org.joda.time.Duration;
import org.joda.time.format.DateTimeFormat;

import java.util.ArrayList;
import java.util.List;

public class ServerConvertImp implements ServerConverter {
    private static final String DAY = "d.";
    private static final String HOUR = "h.";
    private static final String MINUTE = "m.";
    private static final String HOUR_RU = "час";
    private static final String MONTH_RU = "месяц";
    private static final String CURRENCY_RU = "₽";
    private static final String UPTIME = "Uptime";

    private static final String STATUS_NEW = "создание";
    private static final String STATUS_ON = "работает";
    private static final String STATUS_OFF = "выключен";
    private static final String STATUS_EXEC = "выполнение";

    @Override
    public DbServer fromApiToDb(ApiServer apiServer) {
        return new DbServer(
                apiServer.getId(),
                apiServer.getName(),
                apiServer.getMemory(),
                apiServer.getVcpus(),
                apiServer.getDisk(),
                apiServer.getRegion().getName(),
                apiServer.getImage().getId(),
                apiServer.getBackupPriceHourly(),
                apiServer.getBilling().getPaymentDate(),
                apiServer.getBilling().getPaymentAmount(),
                apiServer.getBilling().getPaymentPeriod(),
                apiServer.getBilling().getTotalHours(),
                apiServer.getBilling().getWorkedHours(),
                apiServer.isLocked(),
                apiServer.getStatus(),
                apiServer.getCreatedAt(),
                apiServer.getStartedFirstAt(),
                apiServer.getStartedAt(),
                apiServer.isIsInstall(),
                apiServer.isIsError(),
                apiServer.getPassword(),
                apiServer.getNetworks().getV4().getIpAddress(),
                apiServer.isMbit200()
        );
    }

    @Override
    public Server fromDbToDomain(DbServer dbServer) {
        return new Server(
                dbServer.getId(),
                dbServer.getName(),
                uptimeConvert(dbServer.getStartedAt()),
                statusConverter(dbServer.getStatus())
        );
    }

    @Override
    public Server fromDbToDomainFull(DbServer dbServer) {
        return new Server(
                dbServer.getId(),
                dbServer.getName(),
                dbServer.getMemory(),
                dbServer.getVcpus(),
                dbServer.getDisk(),
                dbServer.getRegion(),
                dbServer.getBackupPriceHourly(),
                nextPaymentDateConvert(dbServer.getPaymentDate()),
                pricePeriodConverter(dbServer.getPaymentAmount(), dbServer.getPaymentPeriod()),
                uptimeConvert(dbServer.getStartedAt()),
                dbServer.isLocked(),
                statusConverter(dbServer.getStatus()),
                dbServer.getCreatedAt(),
                dbServer.getStartedFirstAt(),
                dbServer.getStartedAt(),
                dbServer.isInstall(),
                dbServer.isError(),
                dbServer.getPassword(),
                dbServer.getV4Ip(),
                dbServer.isMbit200()
        );
    }

    @Override
    public List<Server> fromDbToDomainList(List<DbServer> dbList) {
        List<Server> convertList = new ArrayList<>();
        for (DbServer dbServer : dbList) {
            convertList.add(fromDbToDomain(dbServer));
        }
        return convertList;
    }

    @Override
    public List<DbServer> fromApiToDbList(List<ApiServer> apiList) {
        List<DbServer> convertList = new ArrayList<>();
        for (ApiServer apiServer : apiList) {
            convertList.add(fromApiToDb(apiServer));
        }
        return convertList;
    }

    private String uptimeConvert(String date) {

        DateTime dateTime = new DateTime().minus(new DateTime(date).getMillis());
        Duration uptime = new Duration(dateTime.getMillis());

        long upTimeDay = uptime.getStandardDays();
        long upTimeHours = dateTime.getHourOfDay();

        if (upTimeDay > 0) {
            return UPTIME + ": " + upTimeDay + DAY + " " + upTimeHours + HOUR;
        } else if (upTimeHours > 0) {
            return UPTIME + ": " + upTimeHours + HOUR;
        } else {
            return UPTIME + ":" + dateTime.getMinuteOfHour() + MINUTE;
        }
    }

    private String nextPaymentDateConvert(String date) {
        return DateTimeFormat.forPattern("yyyy-MM-dd HH:mm").print(new DateTime(date));
    }

    private String pricePeriodConverter(String price, String period) {
        return priceConverter(price) + "/" + periodConverter(period);
    }

    private String priceConverter(String price) {
        return Double.valueOf(price) + CURRENCY_RU;
    }

    private String periodConverter(String period) {
        switch (period) {
            case "1 HOUR":
                return HOUR_RU;
            case "1 MONTH":
                return MONTH_RU;
            default:
                return "";
        }
    }

    private String statusConverter(String status) {
        switch (status) {
            case "new":
                return STATUS_NEW;
            case "active":
                return STATUS_ON;
            case "off":
                return STATUS_OFF;
            default:
                return STATUS_EXEC;
        }
    }
}

File: app/src/main/java/com/platdmit/simplecloudmanager/domain/converters/implement/ServerConvertImp.kt
package com.platdmit.simplecloudmanager.domain.converters.implement

import com.platdmit.data.api.models.ApiServer
import com.platdmit.data.database.entity.DbServer
import com.platdmit.simplecloudmanager.domain.converters.ServerConverter
import com.platdmit.simplecloudmanager.domain.models.Server
import org.joda.time.DateTime
import org.joda.time.Duration
import org.joda.time.format.DateTimeFormat
import java.util.*

class ServerConvertImp : ServerConverter {
    override fun fromApiToDb(apiServer: ApiServer): DbServer {
        return DbServer(
                apiServer.id.toLong(),
                apiServer.name,
                apiServer.memory,
                apiServer.vcpus,
                apiServer.disk,
                apiServer.region.name,
                apiServer.image.id.toLong(),
                apiServer.backupPriceHourly,
                apiServer.billing.paymentDate,
                apiServer.billing.paymentAmount,
                apiServer.billing.paymentPeriod,
                apiServer.billing.totalHours,
                apiServer.billing.workedHours,
                apiServer.isLocked,
                apiServer.status,
                apiServer.createdAt,
                apiServer.startedFirstAt,
                apiServer.startedAt,
                apiServer.isIsInstall,
                apiServer.isIsError,
                apiServer.password,
                apiServer.networks.v4.ipAddress,
                apiServer.isMbit200
        )
    }

    override fun fromDbToDomain(dbServer: DbServer): Server {
        return Server(
                dbServer.id,
                dbServer.name,
                uptimeConvert(dbServer.startedAt),
                statusConverter(dbServer.status)
        )
    }

    override fun fromDbToDomainFull(dbServer: DbServer): Server {
        return Server(
                dbServer.id,
                dbServer.name,
                dbServer.memory,
                dbServer.vcpus,
                dbServer.disk,
                dbServer.region,
                dbServer.backupPriceHourly,
                nextPaymentDateConvert(dbServer.paymentDate),
                pricePeriodConverter(dbServer.paymentAmount, dbServer.paymentPeriod),
                uptimeConvert(dbServer.startedAt),
                dbServer.isLocked,
                statusConverter(dbServer.status),
                dbServer.createdAt,
                dbServer.startedFirstAt,
                dbServer.startedAt,
                dbServer.isInstall,
                dbServer.isError,
                dbServer.password,
                dbServer.v4Ip,
                dbServer.isMbit200
        )
    }

    override fun fromDbToDomainList(dbList: List<DbServer>): List<Server> = dbList.map { fromDbToDomain(it) }

    override fun fromApiToDbList(apiList: List<ApiServer>): List<DbServer> = apiList.map { fromApiToDb(it) }

    private fun uptimeConvert(date: String): String {
        val dateTime = DateTime().minus(DateTime(date).millis)
        val uptime = Duration(dateTime.millis)
        val upTimeDay = uptime.standardDays
        val upTimeHours = dateTime.hourOfDay.toLong()
        return if (upTimeDay > 0) {
            "$UPTIME: $upTimeDay$DAY $upTimeHours$HOUR"
        } else if (upTimeHours > 0) {
            "$UPTIME: $upTimeHours$HOUR"
        } else {
            UPTIME + ":" + dateTime.minuteOfHour + MINUTE
        }
    }

    private fun nextPaymentDateConvert(date: String): String {
        return DateTimeFormat.forPattern("yyyy-MM-dd HH:mm").print(DateTime(date))
    }

    private fun pricePeriodConverter(price: String, period: String): String {
        return priceConverter(price) + "/" + periodConverter(period)
    }

    private fun priceConverter(price: String): String {
        return java.lang.Double.valueOf(price).toString() + CURRENCY_RU
    }

    private fun periodConverter(period: String): String {
        return when (period) {
            "1 HOUR" -> HOUR_RU
            "1 MONTH" -> MONTH_RU
            else -> ""
        }
    }

    private fun statusConverter(status: String): String {
        return when (status) {
            "new" -> STATUS_NEW
            "active" -> STATUS_ON
            "off" -> STATUS_OFF
            else -> STATUS_EXEC
        }
    }

    companion object {
        private const val DAY = "d."
        private const val HOUR = "h."
        private const val MINUTE = "m."
        private const val HOUR_RU = "час"
        private const val MONTH_RU = "месяц"
        private const val CURRENCY_RU = "₽"
        private const val UPTIME = "Uptime"
        private const val STATUS_NEW = "создание"
        private const val STATUS_ON = "работает"
        private const val STATUS_OFF = "выключен"
        private const val STATUS_EXEC = "выполнение"
    }
}